pkg/tracer/tracer_linux.go

```diff
@@ -479,6 +567,86 @@ func (t *LinuxTracer) getMap(mapName string) (*bpflib.Map, error) {
 	return mp, nil
 }
 
+func (t *LinuxTracer) dispatchPerfEvent(event *common.PerfEvent) {
+	t.tcpConnInsightsLock.Lock()
+	defer t.tcpConnInsightsLock.Unlock()
+
+	if event.HTTPResponse != nil {
+		logger.Tracef("http response: %v", event.HTTPResponse)
+		httpRes := event.HTTPResponse
+
+		conn, ok := t.tcpConnInsights[httpRes.Connection]
+		httpProtocol := "http"
+		if !ok {
+			conn = ConnInsight{
+				ApplicationProtocol: httpProtocol,
+				HttpMetrics:         make(map[HttpCode]*ddsketch.DDSketch),
+			}
+		}
+		conn.ApplicationProtocol = httpProtocol
+
+		latencyCounter, ok := conn.HttpMetrics[httpRes.StatusCode]
+		if !ok {
+			var err error
+			latencyCounter, err = ddsketch.NewDefaultDDSketch(t.config.HttpMetricPrecision)
+			if err != nil {
+				logger.Errorf("can't create dd sketch. Error: %v", err)
+			} else {
+				conn.HttpMetrics[httpRes.StatusCode] = latencyCounter
+				err := latencyCounter.Add(httpRes.ResponseTime.Seconds())
+				if err != nil {
+					logger.Errorf("can't add response time to DDSketch. Error: %v", err)
+				}
+			}
+		}
+		t.tcpConnInsights[httpRes.Connection] = conn
+
+	} else if event.MySQLGreeting != nil {
+		logger.Tracef("mysql greeting: %v", event.MySQLGreeting)
+		mysqlGreeting := event.MySQLGreeting
+
+		conn, ok := t.tcpConnInsights[mysqlGreeting.Connection]
+		if !ok {
+			conn = ConnInsight{
+				ApplicationProtocol: "mysql",
+				HttpMetrics:         make(map[HttpCode]*ddsketch.DDSketch),
+			}
+		}
+		conn.ApplicationProtocol = "mysql"
+		t.tcpConnInsights[mysqlGreeting.Connection] = conn
+	}
+}
+
+func (t *LinuxTracer) enrichTcpConns(conns []common.ConnectionStats) []common.ConnectionStats {
+	logger.Debug("enrich tcp connections")
+	t.tcpConnInsightsLock.Lock()
```

Comment:
You've now locked outside of the loop. There is a fundamental difference with locking inside the loop. If you lock around the loop, nothing now can interleave. First this was possible.

Reply:
This loop should be very quick (no IO required). Probably it's easier/faster locking and unlocking just once.

Reply:
Did you measure? How many connection stats are in the slice typically, maximum

Reply:
It's configured by the properties file. By default it's 2000.

Reply:
We measured: in both cases, total lock and per-iteration lock, it takes pretty same few milliseconds. Updated code with your proposal with function extraction.